core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannel.scala

```diff
@@ -54,41 +87,28 @@ class TransactionMarkerChannel(interBrokerListenerName: ListenerName,
 
   private[transaction]
   def drainQueuedTransactionMarkers(txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker]): Iterable[RequestAndCompletionHandler] = {
-    brokerStateMap.flatMap {case (brokerId: Int, destAndMarkerQueue: DestinationBrokerAndQueuedMarkers) =>
-      val markersToSend: java.util.List[CoordinatorEpochAndMarkers] = new util.ArrayList[CoordinatorEpochAndMarkers] ()
-      destAndMarkerQueue.markersQueue.drainTo (markersToSend)
-      markersToSend.groupBy{ epochAndMarker => (epochAndMarker.metadataPartition, epochAndMarker.coordinatorEpoch) }
-        .map { case((metadataPartition: Int, coordinatorEpoch:Int), buffer: mutable.Buffer[CoordinatorEpochAndMarkers]) =>
-          val txnMarkerEntries = buffer.flatMap{_.txnMarkerEntries }.asJava
-          val requestCompletionHandler = new TransactionMarkerRequestCompletionHandler(
-            this,
-            txnMarkerPurgatory,
-            CoordinatorEpochAndMarkers(metadataPartition, coordinatorEpoch, txnMarkerEntries),
-            brokerId)
-          RequestAndCompletionHandler(destAndMarkerQueue.destBrokerNode, new WriteTxnMarkersRequest.Builder(coordinatorEpoch, txnMarkerEntries), requestCompletionHandler)
-        }
+    brokerStateMap.flatMap {case (brokerId: Int, brokerRequestQueue: BrokerRequestQueue) =>
+      brokerRequestQueue.eachMetadataPartition{ case(partitionId:Int, queue:BlockingQueue[TxnMarkerEntry]) =>
+        val markersToSend: java.util.List[TxnMarkerEntry] = new util.ArrayList[TxnMarkerEntry]()
+        queue.drainTo(markersToSend)
+        val requestCompletionHandler = new TransactionMarkerRequestCompletionHandler(this, txnMarkerPurgatory, partitionId, markersToSend, brokerId)
+        RequestAndCompletionHandler(brokerRequestQueue.node, new WriteTxnMarkersRequest.Builder(markersToSend), requestCompletionHandler)
+      }
     }
   }
 
 
   def addOrUpdateBroker(broker: Node) {
-    if (brokerStateMap.contains(broker.id())) {
-      val brokerQueue = brokerStateMap(broker.id())
-      if (!brokerQueue.destBrokerNode.equals(broker)) {
-        brokerStateMap.put(broker.id(), DestinationBrokerAndQueuedMarkers(broker, brokerQueue.markersQueue))
-        trace(s"Updated destination broker for ${broker.id} from: ${brokerQueue.destBrokerNode} to: $broker")
-      }
-    } else {
-      val markersQueue = new LinkedBlockingQueue[CoordinatorEpochAndMarkers]()
-      brokerStateMap.put(broker.id, DestinationBrokerAndQueuedMarkers(broker, markersQueue))
-      trace(s"Added destination broker ${broker.id}: $broker")
+    brokerStateMap.putIfAbsent(broker.id(), new BrokerRequestQueue(broker)) match {
+      case Some(brokerQueue) => brokerQueue.maybeUpdateNode(broker)
```

Comment:
@dguy @ijuma In that case do we still want to send the queued items to the new broker? I thought that when some brokers are down we will then clear the corresponding queue since the new leader will be on some other brokers. And when it is back it will still be considered as a "new broker" even with the same broker id and it will not automatically be the leader for any partitions until some leader migration happens, in which we will then re-create the queue for this broker and enqueue the requests?